miden-lib/asm/miden/note.masm

```diff
@@ -143,4 +143,328 @@ export.get_serial_number
+export.compute_inputs_hash_cdrop
+    # move number of inputs to the top of the stack 
+    swap
+    # => [num_inputs, inputs_ptr]
+
+    # check that number of inputs is less than 128
+    dup push.128 u32assert2 u32lte assert.err=ERR_NOTE_TOO_MANY_INPUTS
+    # => [num_inputs, inputs_ptr]
+
+    # get the number of double words
+    u32divmod.8 swap
+    # => [num_inputs/8, num_inputs%8, inputs_ptr]
+
+    # get the end_addr for hash_memory_even proc (end address for pairs of words)
+    u32wrapping_mul.2 dup.2 u32wrapping_add movup.2
+    # => [inputs_ptr, end_addr, num_inputs%8]
+
+    # get the padding flag to add it to the capacity part
+    dup.2 push.0 neq 
+    # => [pad_flag, inputs_ptr, end_addr, num_inputs%8]
+    
+    # prepare hasher state for RPO permutation 
+    push.0.0.0 padw padw 
+    # => [C, B, A, inputs_ptr, end_addr, num_inputs%8]
+
+    # hash every pair of words
+    dup.13 dup.13 neq
+    if.true
+        exec.native::hash_memory_even
+    end
+    # => [C', B', A', inputs_ptr', end_addr, num_inputs%8] where inputs_ptr' = end_addr
+
+    # hash remaining input values if there any left
+    # if num_inputs%8 is not ZERO
+    dup.14 push.0 neq
+    if.true
+        # check whether we need to drop more than 4 elements
+
+        # load the remaining double word
+        mem_stream
+        # => [E, D, A', inputs_ptr'+2, end_addr, num_inputs%8]
+
+        # clean the stack
+        movup.12 drop movup.12 drop
+        # => [E, D, A', num_inputs%8]
+
+        # get the number of elements we need to drop
+        # notice that drop_counter could be any number from 1 to 7
+        push.8 movup.13 u32wrapping_sub movdn.12
+        # => [E, D, A', drop_counter]
+
+        ### 0th value ########################################################
+        
+        # if current value is the last value to drop (\"cycle\" number equals to the number of values 
+        # to drop), push 1 instead of 0 to the stack
+        dup.12 eq.1
+        if.true
+            push.1 swap
+        else
+            push.0 swap
+        end
+        # => [e_0, 0/1, e_1, e_2, e_3, d_0, d_1, d_2, d_3, A', drop_counter]
+
+        # we need to drop first value anyway, since number of values is not divisible by 8
+        drop
+        # => [0/1, e_1, e_2, e_3, d_0, d_1, d_2, d_3, A', drop_counter]
+
+        # move the calculated value down the stack 
+        movdn.6
+        # => [e_1, e_2, e_3, d_0, d_1, d_2, 0/1, d_3, A', drop_counter]
+
+        ### 1st value ########################################################
+
+        # prepare the second element of the E Word (e_1) for cdrop instruction
+        # if current value is the last value to drop (\"cycle\" number equals to the number of values 
+        # to drop), push 1 instead of 0 to the stack
+        dup.12 eq.2
+        if.true
+            push.1 swap
+        else
+            push.0 swap
+        end
+        # => [e_1, 0/1, e_2, e_3, d_0, d_1, d_2, 0/1, d_3, A', drop_counter]
```

Comment:
Why do we need `if.true` statement here? shouldn't this be equivalent to just `dup.12 eq.2 swap`?

Reply:
Ah, indeed, we can just use the result of the `eq.n` instruction above instead of this `if`. My bad